testsuite/tests/ast-invariants/test.ml

```diff
@@ -61,14 +61,20 @@ let kind fn =
 
 let rec walk dir =
   Array.iter
-    (fun fn ->
-       if fn = "" || fn.[0] = '.' then
+    (fun base ->
+       if base = "" || base.[0] = '.' then
          ()
        else begin
-         let fn = Filename.concat dir fn in
+         let fn = Filename.concat dir base in
          match kind fn with
          | Other -> ()
-         | Directory -> walk fn
+         | Directory ->
+           (* Don't recurse into any "hidden" directories (.git, etc.).
+              Similarly ignore directories beginning with _ - catches both the
+              ancient use of _ instead of . on Windows and also means we don't
+              recurse into either _ocamltest or _build directories. *)
+           if base.[0] <> '_' && base.[0] <> '.' then
+             walk fn
```

Comment:
I thought we already handled the case `base.[0] = '.'`, so here we should know statically that `base.[0] <> '.'`. What am I missing? Can we put all the blacklisting logic on line 65, with an extra case for `base.[0] = '_'`?

Reply:
I hadn't correctly registered that it was `base.[0] = ` rather than `base = `. The only reason I put it down here was because it's after the call to `stat` and therefore definitely only applies to directories, but that doesn't really matter, I guess.